Implement classification pipeline

bbarnes52 opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## What changes are proposed in this pull request?

This PR introduces a pipeline implementation for building classification models and updates associated utility methods that perform logic conditioned on the template type. Note that the classification pipeline implementation is currently identical to the regression pipeline implementation, this is expected to change as we add additional facilities such as a rebalancing step. A few other notes:
- This PR also adds support for declaring the `positive_class` and threads this value through the mlflow evaluation service.
- This PR updates builtin metrics name which are inconsistent with the metrics supported by the mlflow evaluation service. I _think_ the [docs](https://www.mlflow.org/docs/latest/python_api/mlflow.html) may need to be updated.
- The worst_examples function doesn't work that well, the current error function returns True if the prediction is correct and false otherwise (rather than using the probability of the prediction to compute the error). I think this is OK for the first version in the interest of getting customer feedback ASAP, but please challenge me on this!

Importantly, I have not disabled this behind a feature flag, users will be able to build classification pipelines once this is released. If reviewers would prefer to disable this until we have a template + example repo, I'd be happy to do so.

## How is this patch tested?

Locally + Databricks using the Iris dataset.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

mlflow-automation commented:
<!-- documentation preview -->

### Documentation preview will be available [here](https://output.circle-artifacts.com/output/job/5df11d6c-ba33-4e16-b6bf-4ffbf51e785b/artifacts/0/docs/build/html/index.html).

<details>
<summary>Notes</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated on every commit to this PR.
- Job URL: https://circleci.com/gh/mlflow/mlflow/33342
- Updated at: 2022-10-17 21:58:03.959312

</details>

prithvikannan commented on mlflow/pipelines/utils/metrics.py:
```diff
@@ -41,13 +41,10 @@ def from_custom_metric_dict(cls, custom_metric_dict):
+    PipelineMetric(name="precision_score", greater_is_better=True),
     PipelineMetric(name="f1_score", greater_is_better=True),
     PipelineMetric(name="accuracy_score", greater_is_better=True),
-    PipelineMetric(name="log_loss", greater_is_better=False),
-    PipelineMetric(name="roc_auc", greater_is_better=True),
-    PipelineMetric(name="precision_recall_auc", greater_is_better=True),
```
why do we need to remove this? it doesn't look like we are adding these back either. i'm not following.

bbarnes52 replied:
That's a great question, and one that we should probably follow up on.

When I attempted to use `log_loss` as my primary metric, the train step failed and returned an error that included the list of allowed metric names. The list did not include the three metrics removed here.

dbczumar replied:
Were there other builtin metric names included in the list that aren't captured here? Where was the list of builtins printed from?

bbarnes52 replied:
This is the error thrown by the evaluate step if `log_loss` is included in this list and specified in `pipeline.yaml`:
```
mlflow.exceptions.MlflowException: The metric log_loss is defined in the pipeline's validation criteria but was not returned from mlflow evaluation.
```

Here's a warning from the train step:
```
2022/10/21 12:25:41 WARNING mlflow.pipelines.utils.step: Input metric names {'log_loss'} not found in eval metrics: {'score', 'precision_score', 'false_negatives', 'true_positives', 'accuracy_score', 'recall_score', 'f1_score', 'example_count', 'true_negatives', 'false_positives'}
```

`example_count` and `score` are not captured here, let me know if you'd like to see them added. And also if you'd like me to create a bug report here (unless I've missed something obvious, quite possible!)

prithvikannan reviewed (commented):
This looks awesome @bbarnes52! Very neat solution. I've left a small comment about the metrics changes.

prithvikannan commented:
the PR description says 
> This PR also adds support for declaring the positive_class in the eval step and threads this value through the mlflow evaluation service.

but from what i can tell from reading the code and testing, `positive_class` needs to be declared at the top level. i think the way it currently is makes the most sense, since the positive_class is critical to defining the problem. can you update the PR description to match @bbarnes52?

jinzhang21 commented on mlflow/pipelines/steps/evaluate.py:
```diff
@@ -60,6 +60,12 @@ def _validate_and_apply_step_config(self):
                 "Missing template_name config in pipeline config.",
                 error_code=INVALID_PARAMETER_VALUE,
             )
+        self.positive_class = self.step_config.get("positive_class")
```
Shall we verify that positive_class actually exists (i.e. a valid label) in the dataset[target_col] in the ingest step?

bbarnes52 replied:
Good question, I think it depends. As things stand, the `positive_class` does not need to exist in the raw data, it could be created during the transform step. Do we want to require that `positive_class` exists in the raw data? I don't immediately see a reason to do so, but please challenge me on this!

We could also alternatively verify that `positive_class` exists after the transform step.

Let me know what you think!

jinzhang21 replied:
Let's don't add anything yet given we don't have any user feedback. We can always revisit this decision later on.

dbczumar replied:
As I understand it, the transform step is only used for feature transformations. Transformations to target columns should be performed upstream (e.g. in ETL). The regression pipeline doesn't pass targets to the transform step.

bbarnes52 replied:
`The regression pipeline doesn't pass targets to the transform step, for example.`
-> Right you are, my apologies. Would you like to see this validation performed during ingest (or split?)

dbczumar replied:
No worries! I think ingest is a good place for this.

jinzhang21 reviewed (approved):
LGTM. Do we have a task to follow up with creating an E2E test running through the classification pipeline example repo, once we have it release?

dbczumar commented on mlflow/pipelines/steps/train.py:
```diff
@@ -135,6 +135,12 @@ def _validate_and_apply_step_config(self):
                 error_code=INVALID_PARAMETER_VALUE,
             )
+        self.positive_class = self.step_config.get("positive_class")
+        if not self.positive_class and self.template == "classification/v1":
```
What if `self.positive_class` is `0`, `False`, `None`, or `""`? Perhaps we should check if `positive_class in not in self.step_config` instead.

bbarnes52 replied:
Good call, will update.

dbczumar commented:
> The worst_examples function doesn't work that well, the current error function returns True if the prediction is correct and false otherwise (rather than using the probability of the prediction to compute the error). I think this is OK for the first version in the interest of getting customer feedback ASAP, but please challenge me on this!

Seems okay to me for starters :)

dbczumar commented:
Re: docs updates, once we have a template repo, we should add it / link it here: https://output.circle-artifacts.com/output/job/eb41f865-1ada-416f-af1d-d7c39639852a/artifacts/0/docs/build/html/pipelines.html#pipeline-templates

dbczumar commented on docs/source/python_api/mlflow.pipelines.rst:
```diff
@@ -18,3 +18,12 @@ Regression Pipeline
+
+.. automodule:: mlflow.pipelines.classification.v1.pipeline
+
+.. autoclass:: mlflow.pipelines.classification.v1.pipeline.ClassificationPipeline()
```
Classification pipeline class members don't seem to render in the generated docs (https://output.circle-artifacts.com/output/job/eb41f865-1ada-416f-af1d-d7c39639852a/artifacts/0/docs/build/html/python_api/mlflow.pipelines.html#mlflow.pipelines.classification.v1.pipeline.ClassificationPipeline)

<img width="995" alt="Screen Shot 2022-10-21 at 12 01 43 PM" src="https://user-images.githubusercontent.com/39497902/197270348-dbe0732d-ff9e-4293-a504-54814b074f96.png">

bbarnes52 replied:
Ah nice catch, I didn't override methods that just deferred to their super class. Will update.

dbczumar commented on mlflow/pipelines/classification/v1/pipeline.py:
```diff
@@ -0,0 +1,188 @@
+"""
+.. _mlflow-classification-pipeline:
+
+The MLflow Classification Pipeline is an MLflow Pipeline for developing classification models.
```
Do our docs mention that only binary classification is supported initially? 

Also, what's the behavior if a user passes in a dataset with > 2 classes? Do we throw or build a one-vs-rest classifier? We should probably make sure that we do one of these.

bbarnes52 replied:
Good question. We would built a one-vs-rest classifier. What do you think desired behavior should be here?

dbczumar replied:
One-vs-rest sounds great! Can we make sure to document this behavior?

dbczumar reviewed (commented):
In the design for classification, I see that the following remain listed as goals:

- Split should do stratified sampling.
- Users may specify whether the trained model should output predicted labels or probabilities.
- A declarative API for handling imbalanced data. Class weights should be applied automatically for models which output probabilities. Validation and test metrics should be based on the original distribution. 
- Train step cards should show validation confusion matrices, ROC curve and precision-recall curves.


Are we incorporating all of these?

dbczumar commented:
Ideally, I'd prefer if we introduced the classification template repo (a less visible "test" location is fine for now), introduced test coverage, and then merged this PR once the coverage is in place. If we need to get this PR merged into `master` right now, I won't strongly object.

bbarnes52 commented:
> In the design for classification, I see that the following remain listed as goals:
> 
> * Split should do stratified sampling.
> * Users may specify whether the trained model should output predicted labels or probabilities.
> * A declarative API for handling imbalanced data. Class weights should be applied automatically for models which output probabilities. Validation and test metrics should be based on the original distribution.
> * Train step cards should show validation confusion matrices, ROC curve and precision-recall curves.
> 
> Are we incorporating all of these?

The current plan is to incorporate these after this PR is landed, I don't feel strongly though. @prithvikannan has a PR for the visualizations, the rest would be handled in the next sprint.

> Ideally, I'd prefer if we introduced the classification template repo (a less visible "test" location is fine for now), introduced > test coverage, and then merged this PR once the coverage is in place. If we need to get this PR merged into master sooner > (after comments are addressed), I won't strongly object.

I don't have a strong preference myself. The classification template repo will be nearly identical to the regression template repo, it would not take long to produce (the example repo will diverge a bit).

@jinzhang21 would you like to weigh in on these points?

jinzhang21 commented:
> 
Since all existing tests are passing. I don't object to merging the current PR as is.

On classification template, agreed it shouldn't be too different from the regression one (in fact there are customers adapting the regression template to classification problems already).

However on classification example, I think we should have a small doc to discuss it before implementation. For example,
on the ML problem selection, we should consider
- Ease of understanding the nature of the problem
- Data source (need to vet its license e.g.)
- Extensibility (to future multi-class / imbalanced class problems).
- Should incorporate batch inference

harupy commented on mlflow/pipelines/steps/ingest/__init__.py:
```diff
@@ -77,6 +93,21 @@ def _run(self, output_directory: str) -> BaseCard:
+                cardinality = ingested_df.apply(pd.Series.nunique)[self.target_col]
+                if cardinality != 2:
+                    raise MlflowException(
+                        f"Target column '{self.target_col}' must have a cardinality of 2,"
+                        "found '{cardinality}'.",
```
```suggestion
                        f"Target column '{self.target_col}' must have a cardinality of 2, "
                        f"found '{cardinality}'.",
```

Nit

harupy commented on mlflow/pipelines/steps/ingest/__init__.py:
```diff
@@ -77,6 +93,21 @@ def _run(self, output_directory: str) -> BaseCard:
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+            if self.positive_class is not None:
+                cardinality = ingested_df.apply(pd.Series.nunique)[self.target_col]
```
```suggestion
                cardinality = ingested_df[self.target_col].nunique()
```

Can we only apply `nunique` to the target column?


### Benchmark:

```python
import numpy as np
import pandas as pd
import timeit

df = pd.DataFrame(np.random.randint(0, 10, size=(10000, 5)), columns=list("ABCDE"))


def subscript_then_nunique():
    return df["A"].nunique()


def nunique_then_subscript():
    return df.apply(pd.Series.nunique)["A"]


assert subscript_then_unique() == unique_then_subscript()

print(timeit.timeit(subscript_then_nunique, number=10000), "seconds")
print(timeit.timeit(nunique_then_subscript, number=10000), "seconds")
```

Output

```
0.542079824999746 seconds
4.465884477000145 seconds
```

Commit:
implement classification pipeline

Signed-off-by: Brian Barnes <brian.barnes@databricks.com>
